### PR TITLE
[Config] Add type-hints to public interfaces and classes

### PR DESCRIPTION
**Pull request by jschaedl**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #32179  
| License       | MIT
| Doc PR        | N/A 

This PR adds type hints to all public interfaces of the Config component.

### CODE REVIEW

**Comment by derrabus** [comment]
Please also update those implementations of the updated interfaces that are also part of the Config component.

File: src/Symfony/Component/Config/Loader/LoaderInterface.php [hunk 1]
```diff
@@ -26,7 +26,7 @@ interface LoaderInterface
      *
      * @throws \Exception If something went wrong
      */
-    public function load($resource, $type = null);
+    public function load($resource, ?string $type = null);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Loader/LoaderInterface.php]: `?` should be removed (same below when null is the default), per our CS

File: src/Symfony/Component/Config/Tests/FileLocatorTest.php [hunk 2]
```diff
@@ -115,6 +115,6 @@ public function testLocateEmpty()
     {
         $loader = new FileLocator([__DIR__.'/Fixtures']);
 
-        $loader->locate(null, __DIR__);
+        $loader->locate('', __DIR__);
```
Comment (jschaedl) [inline src/Symfony/Component/Config/Tests/FileLocatorTest.php]: This test was expecting an `InvalidArgumentException` but gets a `TypeError` now.

File: src/Symfony/Component/Config/Loader/LoaderInterface.php (hunk 1 above)
Comment (jschaedl) [reply src/Symfony/Component/Config/Loader/LoaderInterface.php]: done

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jschaedl.

File: src/Symfony/Component/Config/Definition/ArrayNode.php [hunk 3]
```diff
@@ -102,40 +102,32 @@ public function getXmlRemappings()
...
 
     /**
      * Sets if deep merging should occur.
-     *
-     * @param bool $boolean
      */
-    public function setPerformDeepMerging($boolean)
+    public function setPerformDeepMerging(bool $boolean)
     {
         $this->performDeepMerging = (bool) $boolean;
```
Comment (Tobion) [inline src/Symfony/Component/Config/Definition/ArrayNode.php]: forgotten cast

File: src/Symfony/Component/Config/Definition/PrototypeNodeInterface.php [hunk 4]
```diff
@@ -23,5 +23,5 @@ interface PrototypeNodeInterface extends NodeInterface
      *
      * @param string $name The name of the node
```
Comment (Tobion) [inline src/Symfony/Component/Config/Definition/PrototypeNodeInterface.php]: useless doc

File: src/Symfony/Component/Config/FileLocator.php [hunk 5]
```diff
@@ -33,7 +33,7 @@ public function __construct($paths = [])
     /**
      * {@inheritdoc}
      */
-    public function locate($name, $currentPath = null, $first = true)
+    public function locate(string $name, string $currentPath = null, $first = true)
```
Comment (Tobion) [inline src/Symfony/Component/Config/FileLocator.php]: forgotten bool

File: src/Symfony/Component/Config/ConfigCacheFactoryInterface.php [hunk 6]
```diff
@@ -28,5 +28,5 @@ interface ConfigCacheFactoryInterface
      *
      * @return ConfigCacheInterface The cache instance
      */
-    public function cache($file, $callable);
+    public function cache(string $file, $callable);
```
Comment (Tobion) [inline src/Symfony/Component/Config/ConfigCacheFactoryInterface.php]: missing callable type

Comment (jschaedl) [reply src/Symfony/Component/Config/ConfigCacheFactoryInterface.php]: I followed @derrabus advice and skipped the `callable` type. See: https://github.com/symfony/symfony/pull/31984#discussion_r292192250

**Comment by jschaedl** [comment]
@Tobion Thanks. I will create a follow-up PR to address your comments.

**Comment by Tobion** [comment]
@jschaedl I already opened #32286

**Pull request merged**